### ISSUE TITLE
Combining "with_permissions_to" with other named scopes causes unneeded database queries

### ISSUE THREAD

**Issue by ledermann** [stffn/declarative_authorization#7]
There is a critical issue in d_a which fires up unneeded database queries. I stumbled upon the issue while profiling my own application, which makes heavy use of d_a AND uses a lot of named scopes. Some requests in my app are very slow and I discovered that there are many SQL queries in the log which are not needed. This queries are generated by d_a. Because I don't know how to fix it I will give you a detailed description to reproduce the issue based on the d_a_demo_app:

1) Install the official demo app from http://github.com/stffn/decl_auth_demo_app on your development machine

2) The demo app does not use any named scope, so add this example scope to talk.rb:

```
named_scope :with_abstract, :conditions => "talks.abstract IS NOT NULL"
```

3) Start the console and type in this example query:

```
Conference.first.talks.with_abstract.with_permissions_to(:read).find :all, :limit => 2
```

4) In the development.log you find this:

```
SELECT * FROM "conferences" LIMIT 1
SELECT * FROM "talks" WHERE ("talks".conference_id = 1) AND ((talks.abstract IS NOT NULL) AND ("talks".conference_id = 1)) 
SELECT "talks".* FROM "talks" INNER JOIN "conferences" ON "conferences".id = "talks".conference_id WHERE ("talks".conference_id = 1) AND (((("conferences"."published" = 't')) AND (talks.abstract IS NOT NULL)) AND ("talks".conference_id = 1)) LIMIT 2
```

Please have a look at the second query. It is not needed at all and should not be there! So far as I see, this time consuming query is only generated if the scope "with_permissions_to" is combined with an existing named scope.

You also see a second issue (duplicate condition the WHERE clause), but this is another story (not as important).

**Comment by stffn** [comment]
Let's start with the second issue: as long as there is no measurement showing any performance issues with the duplicate conditions, I believe in the DBM's query optimizer to handle them.

**Comment by stffn** [comment]
Fixed by e90082c655d51d785bb01a3f6262755df824bb95

**Comment by ledermann** [comment]
To the second issue: Full agree.

But to the main issue: It seems to me that your commit does not fix the bug. Do you have tested it with my description above? I have checked it with your commit and the bug still occurs for me. 
Sorry for not providing a failing test, I don't know how to build it in this case.

**Comment by ledermann** [comment]
One more hint: I found that the bug was introduced by 824b11bb80ead238e8991d46f169e06567f2adb7

**Comment by stffn** [comment]
I tested as you described and could reproduce it before e90082c, but it works for me now.  If there is a similar bug because of 824b11b, it isn't reproduced in your described test.

**Comment by ledermann** [comment]
Strange. Did you notice that the submodule in the current d_a_demo is not up to date? Here a detailed bash script to reproduce the bug (based on my description above):

```
git clone git://github.com/stffn/decl_auth_demo_app.git
cd decl_auth_demo_app/
cp config/database.yml.example config/database.yml
cp db/development.sqlite3.example db/development.sqlite3
git submodule init
git submodule update
cd vendor/plugins/declarative_authorization/
git checkout master
cd -
script/console
Talk.named_scope :with_abstract, :conditions => "talks.abstract IS NOT NULL"
Conference.first.talks.with_abstract.with_permissions_to(:read).find :all, :limit => 2
exit
cat log/development.log
```

Just run this script on your local (OS X) machine and you will see the unnecessary query:

```
SELECT * FROM "talks" WHERE ("talks".conference_id = 1)
```

Hope this helps.

**Comment by ledermann** [comment]
Now this example is fixed by 99021e4ee06eac55daf745e37277674485cf471d. Thank you very much! But sadly I'm still having issues (since 824b11bb80ead238e8991d46f169e06567f2adb7) with unnecessary queries in my own app causing serious performance problems. I will try to extract them...

**Comment by ledermann** [comment]
Fixed by bcfbbce9828d68297b90a342c535b6bc0c20e80a. Now there are no more unnecessary queries. Thanks!
